File: dat-pypi/src/saro_dat/dat_cms_manager.py
import http.client
import logging
import socket
import threading
import time
import urllib.error
import urllib.request
from typing import Optional, Union
from urllib.parse import urlparse

from .dat_manager import DatManager
from .dat import Dat, DatPayload
from . import error as E
from .error import DatError, DatRetry
from .util import parse_u64

logger = logging.getLogger(__name__)


class _CleanupTimer(threading.Timer):
    def run(self):
        try:
            super().run()
        finally:
            self.function = None
            self.args = None
            self.kwargs = None


class _SameOriginRedirect(urllib.request.HTTPRedirectHandler):
    def __init__(self, origin: str):
        super().__init__()
        self._origin = origin

    def redirect_request(self, req, fp, code, msg, headers, newurl):
        parsed = urlparse(newurl)
        if (parsed.scheme, parsed.netloc) != self._origin:
            raise urllib.error.URLError("cross-origin redirect is not allowed")
        return super().redirect_request(req, fp, code, msg, headers, newurl)


def _http_status_error(status: int) -> DatError:
    if status == 401:
        return DatError(E.CMS_UNAUTHORIZED, "http 401")
    if status == 403:
        return DatError(E.CMS_FORBIDDEN, "http 403")
    if status == 404:
        return DatError(E.CMS_ENDPOINT_NOT_FOUND, "http 404")
    if 500 <= status <= 599:
        return DatError(E.CMS_SERVER_ERROR, f"http {status}")
    return DatError(E.CMS_HTTP_STATUS, f"http {status}")

class DatCmsManager:
    DAT_CMS_API_VERSION = "v1"

    def __init__(
        self,
        uri: str,
        token: str,
        interval_seconds: int = 60,
        verify_only: bool = False,
        dat_manager: Optional[DatManager] = None,
        connect_timeout_seconds: float = 5,
        sync_timeout_seconds: float = 15,
    ):
        self._uri = uri
        self._token = token
        self._interval_seconds = interval_seconds
        self._verify_only = verify_only
        self._manager = dat_manager or DatManager()
        self._connect_timeout_seconds = connect_timeout_seconds
        self._sync_timeout_seconds = sync_timeout_seconds
        self._version = 0
        self._state_lock = threading.Lock()
        self._sync_lock = threading.Lock()
        self._timer: Optional[threading.Timer] = None
        self._stopped = False
        self._last_error: Optional[DatError] = DatError(E.CMS_NOT_SYNCED)

        self.sync()

        if self._interval_seconds > 0:
            self._schedule_sync()

    def _schedule_sync(self):
        with self._state_lock:
            if not self._stopped:
                self._timer = _CleanupTimer(self._interval_seconds, self._run_sync_task)
                self._timer.daemon = True
                self._timer.start()

    def _run_sync_task(self):
        try:
            self.sync()
        finally:
            self._schedule_sync()

    def stop(self):
        with self._state_lock:
            self._stopped = True
            if self._timer:
                self._timer.cancel()
                self._timer = None

    def last_error(self) -> Optional[DatError]:
        return self._last_error

    def sync(self):
        try:
            self.sync_or_raise()
            self._last_error = None
        except DatError as e:
            if e.retry is not DatRetry.STATE:
                self._last_error = e
                logger.error("[CRITICAL] DAT CMS SYNC %s: %s", self._uri, e)
        except Exception as e:
            self._last_error = DatError(E.CMS_UNKNOWN, "unclassified cms failure", e)
            logger.exception("[CRITICAL] DAT CMS SYNC %s", self._uri)

    def sync_or_raise(self):
        if not self._sync_lock.acquire(blocking=False):
            logger.debug("cms sync skipped, previous sync still running: %s", self._uri)
            raise DatError(E.CMS_SYNC_IN_PROGRESS)

        try:
            url = f"{self._uri}?version={self._version}"

            request = urllib.request.Request(url)
            request.add_header("Authorization", self._token)

            try:
                parsed = urlparse(self._uri)
                opener = urllib.request.build_opener(_SameOriginRedirect((parsed.scheme, parsed.netloc)))
                timeout = self._sync_timeout_seconds or self._connect_timeout_seconds or None
                response_cm = opener.open(request, timeout=timeout)
            except urllib.error.HTTPError as e:
                raise _http_status_error(e.code) from e
            except urllib.error.URLError as e:
                raise DatError(E.CMS_UNREACHABLE, f"cannot reach {self._uri}", e) from e

            with response_cm as response:
                if not (200 <= response.status <= 299):
                    raise _http_status_error(response.status)

                try:
                    body_bytes = response.read()
                except (OSError, socket.timeout, http.client.HTTPException) as e:
                    raise DatError(E.CMS_UNREACHABLE, f"cannot read {self._uri}", e) from e
                if not body_bytes:
                    raise DatError(E.CMS_MALFORMED, "response is empty")
                if any(byte > 0x7f for byte in body_bytes):
                    raise DatError(E.CMS_MALFORMED, "response is not strict ASCII")
                body = body_bytes.decode('ascii')
                lines = body.split('\n', 1)
                new_version_str = lines[0]
                new_certificates = lines[1].strip() if len(lines) == 2 else ""

                if not new_version_str:
                    raise DatError(E.CMS_MALFORMED, "version line is empty")

                try:
                    if not new_version_str.isascii() or not new_version_str.isdecimal() or not new_version_str:
                        raise DatError(E.CMS_MALFORMED, "version line is not a plain decimal u64")
                    new_version = parse_u64(new_version_str)
                except DatError as e:
                    raise DatError(E.CMS_MALFORMED, "version line is not a plain decimal u64", e) from e

                if not new_certificates:
                    logger.debug("No new certificate: %s", url)
                    return

                if new_version < self._version:
                    logger.warning("%s: %d -> %d", E.CMS_VERSION_RESET, self._version, new_version)

                try:
                    renew_count = self._manager.imports(new_certificates, False)
                except DatError as e:
                    raise DatError(E.CMS_IMPORT_FAILED, "cannot apply received certificates", e) from e

                self._version = new_version
                logger.debug("Renewed %d certificates for version %d: %s", renew_count, new_version, url)
        finally:
            self._sync_lock.release()

    def get_manager(self) -> DatManager:
        return self._manager

    def issue(self, plain: Union[bytes, str, None], secure: Union[bytes, str, None]) -> str:
        return self._manager.issue(plain, secure)

    def parse(self, dat: Union[Dat, str, None]) -> DatPayload:
        return self._manager.parse(dat)

    @classmethod
    def builder(cls):
        return DatCmsManagerBuilder()

class DatCmsManagerBuilder:
    def __init__(self):
        self._uri = "http://localhost:8088"
        self._token = ""
        self._verify_only = False
        self._interval_seconds = 60
        self._connect_timeout_seconds = 5
        self._sync_timeout_seconds = 15

    def uri(self, uri: str):
        self._uri = uri.rstrip('/')
        return self

    def token(self, token: str):
        self._token = token
        return self

    def verify_only(self, verify_only: bool):
        self._verify_only = verify_only
        return self

    def interval_seconds(self, interval_seconds: int):
        self._interval_seconds = interval_seconds
        return self

    def interval_off(self):
        return self.interval_seconds(0)

    def connect_timeout_seconds(self, seconds: float):
        self._connect_timeout_seconds = seconds
        return self

    def sync_timeout_seconds(self, seconds: float):
        self._sync_timeout_seconds = seconds
        return self

    def build(self) -> DatCmsManager:
        from urllib.parse import urlparse
        parsed = urlparse(self._uri)
        
        if parsed.scheme not in ('http', 'https'):
            raise DatError(E.CONFIG_URI_INVALID, "scheme must be http or https")
        if parsed.path and parsed.path != '/':
            raise DatError(E.CONFIG_URI_INVALID, f"must be path-less: {self._uri}")
        if parsed.query:
            raise DatError(E.CONFIG_URI_INVALID, f"must be query-less: {self._uri}")

        path = "/v1/certs/verify-only" if self._verify_only else "/v1/certs"

        final_uri = f"{parsed.scheme}://{parsed.netloc}{path}"
        
        return DatCmsManager(
            uri=final_uri,
            token=self._token,
            interval_seconds=self._interval_seconds,
            verify_only=self._verify_only,
            connect_timeout_seconds=self._connect_timeout_seconds,
            sync_timeout_seconds=self._sync_timeout_seconds,
        )

File: dat-ruby/lib/saro/dat/dat_cms_manager.rb
# frozen_string_literal: true

require 'net/http'
require 'uri'
require 'logger'
require 'thread'
require_relative 'error'
require_relative 'dat_manager'
require_relative 'dat'

module Saro
  module Dat
    class DatCmsManager
      DAT_CMS_API_VERSION = "v1"

      STOP_JOIN_TIMEOUT_SECONDS = 1.0

      def initialize(uri:, token:, interval_seconds: 60, verify_only: false, dat_manager: nil,
                     connect_timeout_seconds: 5, sync_timeout_seconds: 15)
        @uri = uri
        @token = token
        @interval_seconds = interval_seconds
        @verify_only = verify_only
        @manager = dat_manager || DatManager.new
        @connect_timeout_seconds = connect_timeout_seconds
        @sync_timeout_seconds = sync_timeout_seconds
        @active_http = nil
        @version = 0
        @lock = Mutex.new
        @lifecycle = Mutex.new
        @stop_cond = ConditionVariable.new
        @stopped = false
        @logger = Logger.new($stdout)
        @logger.level = Logger::DEBUG
        @last_error = Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_NOT_SYNCED)

        sync

        if @interval_seconds > 0
          schedule_sync
        end
      end

      def stop
        thread = nil
        @lifecycle.synchronize do
          return if @stopped
          @stopped = true
          @stop_cond.broadcast
          thread = @thread
          @active_http&.finish if @active_http&.started?
        end
        thread&.join(STOP_JOIN_TIMEOUT_SECONDS)
        nil
      end

      def stopped?
        @lifecycle.synchronize { @stopped }
      end

      attr_reader :last_error

      def version
        @version
      end

      def sync
        err = sync_or_raise
        @last_error = nil
        err
      rescue Saro::Dat::Error => e
        unless e.retry == :state
          @last_error = e
          @logger.error("[CRITICAL] DAT CMS SYNC #{@uri}: #{e.code} #{e.detail}")
        end
        nil
      rescue StandardError => e
        @last_error = Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_UNKNOWN, "unclassified cms failure", cause: e)
        @logger.error("[CRITICAL] DAT CMS SYNC #{@uri}: #{e.message}")
        nil
      end

      def sync_or_raise
        unless @lock.try_lock
          @logger.debug("cms sync skipped, previous sync still running: #{@uri}")
          raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_SYNC_IN_PROGRESS)
        end

        begin
          url = URI("#{@uri}?version=#{@version}")
          request = Net::HTTP::Get.new(url)
          request["Authorization"] = @token

          response =
            begin
              Net::HTTP.start(url.host, url.port, use_ssl: url.scheme == 'https',
                              open_timeout: (@connect_timeout_seconds.zero? ? nil : @connect_timeout_seconds),
                              read_timeout: (@sync_timeout_seconds.zero? ? nil : @sync_timeout_seconds)) do |http|
                @lifecycle.synchronize { @active_http = http }
                http.request(request)
              ensure
                @lifecycle.synchronize { @active_http = nil }
              end
            rescue StandardError => e
              raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_UNREACHABLE, "cannot reach #{@uri}", cause: e)
            end

          status = response.code.to_i
          raise self.class.http_status_error(status) unless status.between?(200, 299)

          body = response.body
          raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_MALFORMED, "response is empty") if body.nil? || body.empty?
          unless body.ascii_only?
            raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_MALFORMED, "response is not strict ASCII")
          end

          lines = body.split("\n", 2)
          new_version_str = lines[0]
          new_certificates = lines.length == 2 ? lines[1].strip : ""

          unless new_version_str.match?(/\A[0-9]+\z/)
            raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_MALFORMED, "version line is not a plain decimal integer")
          end
          new_version = new_version_str.to_i

          if new_certificates.empty?
            @logger.debug("No new certificate: #{url}")
            return nil
          end

          if new_version < @version
            @logger.warn("#{Saro::Dat::ErrorCode::CMS_VERSION_RESET}: #{@version} -> #{new_version}")
          end

          renew_count =
            begin
              @manager.imports(new_certificates, clear: false)
            rescue Saro::Dat::Error => e
              raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_IMPORT_FAILED, "cannot apply received certificates", cause: e)
            end

          @version = new_version
          @logger.debug("Renewed #{renew_count} certificates for version #{new_version}: #{url}")
          nil
        ensure
          @lock.unlock
        end
      end

      def self.http_status_error(status)
        case status
        when 401 then Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_UNAUTHORIZED, "http 401")
        when 403 then Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_FORBIDDEN, "http 403")
        when 404 then Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_ENDPOINT_NOT_FOUND, "http 404")
        when 500..599 then Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_SERVER_ERROR, "http #{status}")
        else Saro::Dat::Error.new(Saro::Dat::ErrorCode::CMS_HTTP_STATUS, "http #{status}")
        end
      end

      def get_manager
        @manager
      end

      def issue(plain, secure)
        @manager.issue(plain, secure)
      end

      def parse(dat)
        @manager.parse(dat)
      end

      def self.builder
        DatCmsManagerBuilder.new
      end

      private

      def schedule_sync
        @thread = Thread.new do
          loop do
            break unless wait_interval
            run_sync_task
          end
        end
      end

      def wait_interval
        @lifecycle.synchronize do
          return false if @stopped
          @stop_cond.wait(@lifecycle, @interval_seconds)
          !@stopped
        end
      end

      def run_sync_task
        sync
      rescue StandardError => e
        @logger.error("Error in sync task: #{e.message}")
      end
    end

    class DatCmsManagerBuilder
      def initialize
        @uri = "http://localhost:8088"
        @token = ""
        @verify_only = false
        @interval_seconds = 60
      end

      def uri(uri)
        @uri = uri.delete_suffix('/')
        self
      end

      def token(token)
        @token = token
        self
      end

      def verify_only(verify_only)
        @verify_only = verify_only
        self
      end

      def interval_seconds(interval_seconds)
        @interval_seconds = interval_seconds
        self
      end

      def interval_off
        @interval_seconds = 0
        self
      end

      def connect_timeout_seconds(seconds)
        @connect_timeout_seconds = seconds
        self
      end

      def sync_timeout_seconds(seconds)
        @sync_timeout_seconds = seconds
        self
      end

      def build
        parsed =
          begin
            URI.parse(@uri)
          rescue URI::InvalidURIError => e
            raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CONFIG_URI_INVALID, "cannot be parsed as a uri", cause: e)
          end

        unless %w[http https].include?(parsed.scheme)
          raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CONFIG_URI_INVALID, "scheme must be http or https")
        end
        if parsed.path && parsed.path != '' && parsed.path != '/'
          raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CONFIG_URI_INVALID, "must be path-less: #{@uri}")
        end
        if parsed.query
          raise Saro::Dat::Error.new(Saro::Dat::ErrorCode::CONFIG_URI_INVALID, "must be query-less: #{@uri}")
        end

        path = @verify_only ? "/v1/certs/verify-only" : "/v1/certs"
        final_uri = "#{parsed.scheme}://#{parsed.host}:#{parsed.port}#{path}"

        DatCmsManager.new(
          uri: final_uri,
          token: @token,
          interval_seconds: @interval_seconds,
          verify_only: @verify_only,
          connect_timeout_seconds: @connect_timeout_seconds,
          sync_timeout_seconds: @sync_timeout_seconds
        )
      end
    end
  end
end
